### PR TITLE
UNR-1375: Save deduplicated schema names in schema db

### PR DESCRIPTION
**Pull request by mattyoung-improbable**
#### Description
- Keep track of generated schema name for actors saved in SchemaDB.
- Use this info to keep names stable and track collisions.
- Update collisions warning to be clearer.

#### Release note
- Fixed an issue with schema name collisions.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SchemaDatabase.h [hunk 1]
```diff
@@ -27,6 +27,9 @@ struct FSchemaData
 {
 	GENERATED_USTRUCT_BODY()
 
+	UPROPERTY(VisibleAnywhere)
+	FString GeneratedSchemaName;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SchemaDatabase.h]: Can this be EDITOR_ONLY_DATA?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 2]
```diff
@@ -37,20 +37,31 @@ DEFINE_LOG_CATEGORY(LogSpatialGDKSchemaGenerator);
...
-TMap<UClass*, FString> ClassToSchemaName;
-TMap<FString, UClass*> UsedSchemaNames;
+TMap<FString, FString> ClassPathToSchemaName;
+TMap<FString, FString> SchemaNameToClassPath;
+TMap<FString, TSet<FString>> PotentialSchemaNameCollisions;
 
 namespace
 {
 
+void AddPotentialNameCollision(FString DesiredSchemaName, FString ClassPath, FString GeneratedSchemaName)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: All these args can be const refs

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 3]
```diff
@@ -37,20 +37,31 @@ DEFINE_LOG_CATEGORY(LogSpatialGDKSchemaGenerator);
...
+{
+	if (!PotentialSchemaNameCollisions.Contains(DesiredSchemaName))
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: How about a PotentialSchemaNameCollisions.FindOrAdd(...).Add(...) ?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp [hunk 4]
```diff
@@ -78,6 +78,11 @@ bool FSpatialGDKEditor::GenerateSchema(bool bFullScan)
 
 	TryLoadExistingSchemaDatabase();
 
+	if (bFullScan)
+	{
+		DeleteGeneratedSchemaFiles();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: Please add a comment about why this is necessary

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SchemaDatabase.h (hunk 1 above)
Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SchemaDatabase.h]: Potentially. Not sure how non-editor commandlets behave with that, and shouldn't be too much of a burden yet.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp (hunk 2 above)
Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Done

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 5]
```diff
@@ -204,24 +210,46 @@ bool ValidateIdentifierNames(TArray<TSharedPtr<FUnrealType>>& TypeInfos)
...
-		int Suffix = 0;
-		while (UsedSchemaNames.Contains(SchemaName))
+		FString DesiredSchemaName = SchemaName;
+
+		if (ClassPathToSchemaName.Contains(ClassPath))
 		{
-			UE_LOG(LogSpatialGDKSchemaGenerator, Warning, TEXT("Class name collision after removing non-alphanumeric characters. Name '%s' collides for types '%s' and '%s'. Will add a suffix to resolve the collision."),
-				*SchemaName, *Class->GetPathName(), *UsedSchemaNames[SchemaName]->GetPathName());
+			// Class already has generated schema name.
+			AddPotentialNameCollision(DesiredSchemaName, ClassPath, ClassPathToSchemaName[ClassPath]);
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Wouldn't we have already added it in `ResetUsedNames`?

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp (hunk 3 above)
Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Done

File: SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp (hunk 4 above)
Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp]: It isn't, removed.

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp [hunk 6]
```diff
@@ -204,24 +210,46 @@ bool ValidateIdentifierNames(TArray<TSharedPtr<FUnrealType>>& TypeInfos)
...
+			AddPotentialNameCollision(DesiredSchemaName, ClassPath, SchemaName);
+		}
+		AddPotentialNameCollision(SchemaName, ClassPath, SchemaName);
+	}
+
+	for (const auto& Collision : PotentialSchemaNameCollisions)
+	{
+		if (Collision.Value.Num() > 1)
+		{
+			UE_LOG(LogSpatialGDKSchemaGenerator, Warning, TEXT("Class name collision after removing non-alphanumeric characters. Name '%s' collides for types [%s]"),
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Nit
```suggestion
			UE_LOG(LogSpatialGDKSchemaGenerator, Warning, TEXT("Class name collision after removing non-alphanumeric characters. Name '%s' collides for classes [%s]"),
```

Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Done

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp (hunk 5 above)
Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp]: Yes, Done

**Pull request merged**